src/osdc/Objecter.h

```diff
@@ -527,15 +527,15 @@ struct ObjectOperation {
 	  obj_list_watch_response_t resp;
 	  decode(resp, p);
 	  if (pwatchers) {
-	    for (list<watch_item_t>::iterator i = resp.entries.begin() ;
-		 i != resp.entries.end() ; ++i) {
+	    for (const auto& watch_item : resp.entries) {
 	      obj_watch_t ow;
-	      string sa = i->addr.get_legacy_str();
-	      strncpy(ow.addr, sa.c_str(), 256);
-	      ow.watcher_id = i->name.num();
-	      ow.cookie = i->cookie;
-	      ow.timeout_seconds = i->timeout_seconds;
-	      pwatchers->push_back(ow);
+	      string sa = watch_item.addr.get_legacy_str();
+	      strncpy(ow.addr, sa.c_str(), sizeof(ow.addr) - 1);
+	      ow.addr[sizeof(ow.addr) - 1] = '\0';
```

Comment:
Does this assume that ow.addr is full of '\0' bytes?  If so, this line isn't needed because it will be set already because the strncpy() didn't use the last byte.  If not, then you need to put the 0 byte at the end what strncpy() transferred.

size_t len = strlen(sa.c_str());
len = std::max(len, sizeof(ow.addr) - 1);
ow.addr[len] = '\0';

Reply:
It won't be automatically constructed with '\0' bytes filled.

Reply:
> Does this assume that ow.addr is full of '\0' bytes?

no. `ow` is allocated on stack. and `ow.addr` is not zero-initialized. i think the issue here is that `strncpy()` cannot ensure the dest string is nul terminated. i don't really want to add an extra call of `strlen()` and `max()`. for they render the code more complicated, and also take extra CPU cycles at runtime.

- as in the best case, `sa.size()` is less than `sizeof(ow.addr) - 1`, and since `sa.c_str()` is already nul terminated, `strncpy(ow.addr, sa.c_str(), sizeof(ow.addr))` will also copy the ending nul. it does not hurt to set the nul at the end of buffer.
- otherwise, `sa.size()` is greater or equal to `sizeof(ow.addr) - 1`, `strncpy(ow.addr, sa.c_str(), sizeof(ow.addr))` will fail to copy the ending nul, it would be important to set the terminating nul.

Reply:
@tchaikov I forgot that strncpy would write the nul byte if it could.